deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/DeepLinkProcessor.kt

```diff
@@ -0,0 +1,580 @@
+     * return a string which contains the calls to those methods.
+     *
+     *
+     * e.g. "method1(), method2()" etc.
+     *
+     * @param urisTrie                The [UrlTree] containing all Urls that can be matched.
+     * @param deepLinkRegistryBuilder The builder used to add the methods
+     * @return
+     */
+    private fun getStringMethodNames(
+        urisTrie: Root,
+        deepLinkRegistryBuilder: TypeSpec.Builder
+    ): StringBuilder {
+        val stringMethodNames = StringBuilder()
+        for ((i, charSequence: CharSequence) in urisTrie.getStrings().withIndex()) {
+            val methodName = "matchIndex$i"
+            stringMethodNames.append(methodName).append("(), ")
+            deepLinkRegistryBuilder.addMethod(
+                MethodSpec.methodBuilder(methodName)
+                    .addModifiers(Modifier.PRIVATE, Modifier.STATIC)
+                    .returns(String::class.java)
+                    .addCode(CodeBlock.builder().add("return \$S;", charSequence).build()).build()
+            )
+        }
+        return stringMethodNames
+    }
+
+    private class IncrementalMetadata(val incremental: Boolean, val customAnnotations: Set<XTypeElement>)
+
+    companion object {
+        private const val PACKAGE_NAME = "com.airbnb.deeplinkdispatch"
+        private const val OPTION_CUSTOM_ANNOTATIONS = "deepLink.customAnnotations"
+        private const val OPTION_INCREMENTAL = "deepLink.incremental"
+        private val CLASS_BASE_DEEP_LINK_DELEGATE =
+            ClassName.get(PACKAGE_NAME, "BaseDeepLinkDelegate")
+        private val CLASS_UTILS = ClassName.get(Utils::class.java)
+        private val DEEP_LINK_CLASS = DeepLink::class
+        private val DEEP_LINK_SPEC_CLASS = DeepLinkSpec::class
+        const val REGISTRY_CLASS_SUFFIX = "Registry"
+
+        /**
+         * For the given element find all custom deeplink elements on it and build all possible
+         * URIs that are supported by theset custom deeplinks.
+         */
+        private fun getAllDeeplinkUrIsFromCustomDeepLinksOnElement(
+            element: XElement,
+            prefixesMap: Map<XType, Array<String>>
+        ): List<String> {
+            return element.annotatedAnnotation(DEEP_LINK_SPEC_CLASS).flatMap { customAnnotation ->
+                val suffixes = customAnnotation.get<List<String>>("value")
+                val prefixes = prefixesMap[customAnnotation.type]
+                    ?: throw DeepLinkProcessorException(
+                        "Unable to find annotation '${customAnnotation.qualifiedName}' you must " +
+                            "update 'deepLink.customAnnotations' within the build.gradle"
+                    )
+                prefixes.flatMap { prefix -> suffixes.map { suffix -> prefix + suffix } }
+            }
+        }
+
+        private fun XElement.annotatedAnnotation(klass: KClass<*>): List<XAnnotation> =
```

Comment:
Looks like this could be reified function version as we only call this with classes that are on the processor classpath

```
internal inline fun <reified T : Annotation> XElement.findAnnotatedAnnotation(): List<XAnnotation> {
    return getAllAnnotations().filter { annotation -> annotation.type.typeElement?.hasAnnotation(T::class) == true }
}
```